components/client/src/main/java/com/hotels/styx/client/connectionpool/TrackedConnectionDecorator.java

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright (C) 2013-2018 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.styx.client.connectionpool;
+
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Ticker;
+import com.hotels.styx.api.HttpRequest;
+import com.hotels.styx.api.HttpResponse;
+import com.hotels.styx.api.client.Connection;
+import com.hotels.styx.api.client.Origin;
+import rx.Observable;
+
+import java.util.function.Supplier;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Provides wrapper for connection, that stores a connection expiration time. Also provides a method for verification of
+ * a time pasted.
+ */
+class TrackedConnectionDecorator implements Connection {
+    private final Connection nettyConnection;
+    private final long connectionExpirationSeconds;
+    private final Stopwatch stopwatch;
+
+    TrackedConnectionDecorator(Connection nettyConnection, long connectionExpirationSeconds, Supplier<Ticker> tickerSupplier) {
+        this.connectionExpirationSeconds = connectionExpirationSeconds;
+        this.nettyConnection = nettyConnection;
```

Comment:
Please check parameters for nulls as this prevents NullPointerExceptions at runtime.